plugins/inputs/tencentcloudcm/tencentcloudcm.go

```diff
@@ -0,0 +1,448 @@
+				if region.RegionName == "" {
+					return fmt.Errorf("region is empty")
+				}
+				_, ok := Registry[namespace.Name]
+				if len(region.Instances) == 0 && !ok {
+					return fmt.Errorf(
+						"unsupported namespace %s for discovering instances, please specify instances and dimensions",
+						namespace.Name,
+					)
+				}
+			}
+		}
+	}
+
+	// start discovery
+	t.discoverTool = NewDiscoverTool(t.Log)
+	go t.discoverTool.Discover(t.Accounts, t.DiscoveryInterval, t.Endpoint)
+	t.discoverTool.DiscoverMetrics()
+	for i := range t.Accounts {
+		for j := range t.Accounts[i].Namespaces {
+			// use discovered metrics if not specified
+			if len(t.Accounts[i].Namespaces[j].Metrics) == 0 {
+				metrics, ok := t.discoverTool.DiscoveredMetrics[t.Accounts[i].Namespaces[j].Name]
+				if !ok {
+					errorInfo := fmt.Sprintf("unsupported namespace %s for discovering metrics, please specify metrics",
+						t.Accounts[i].Namespaces[j].Name)
+					t.Log.Error(errorInfo)
+					return errors.New(errorInfo)
+				}
+				t.Accounts[i].Namespaces[j].Metrics = metrics
+			}
+		}
+	}
+
+	t.client = &cloudmonitorClient{Accounts: t.Accounts}
+
+	return nil
+
+}
+
+func (t *TencentCloudCM) updateWindow(relativeTo time.Time) {
+	windowEnd := relativeTo.Add(-time.Duration(t.Delay))
+
+	t.windowStart = windowEnd.Add(-time.Duration(t.Period) * 2)
+
+	t.windowEnd = windowEnd
+}
+
+// GetMetricObjects gets all metric objects
+func (t *TencentCloudCM) GetMetricObjects() []MetricObject {
+
+	// metricObejcts holds all metrics with it's corresponding region,
+	// namespace, credential and instances(dimensions) information.
+	metricObjects := []MetricObject{}
+
+	// construct metric object
+	for i := range t.Accounts {
+		for j := range t.Accounts[i].Namespaces {
+			for k := range t.Accounts[i].Namespaces[j].Regions {
+				for l := range t.Accounts[i].Namespaces[j].Metrics {
+					instances := t.Accounts[i].Namespaces[j].Regions[k].Instances
+					if len(instances) == 0 {
+						instances = t.discoverTool.GetInstances(t.Accounts[i].Name, t.Accounts[i].Namespaces[j].Name, t.Accounts[i].Namespaces[j].Regions[k].RegionName)
+					}
+					if len(instances) == 0 {
+						continue
+					}
+					metricObjects = append(metricObjects, MetricObject{
+						t.Accounts[i].Namespaces[j].Metrics[l],
+						t.Accounts[i].Namespaces[j].Regions[k].RegionName,
+						t.Accounts[i].Namespaces[j].Name,
+						t.Accounts[i],
+						instances,
+					})
+				}
+			}
+		}
+	}
+	return metricObjects
+}
+
+func (t *TencentCloudCM) Gather(acc telegraf.Accumulator) error {
+
+	t.updateWindow(time.Now())
+
+	metricObjects := t.client.GetMetricObjects(*t)
+
+	lmtr := limiter.NewRateLimiter(t.RateLimit, time.Second)
+	defer lmtr.Stop()
+
+	wg := sync.WaitGroup{}
+	rLock := sync.Mutex{}
+	results := []GetMonitorDataResponse{}
+
+	// requestIDMap contains request ID and metric objects for later aggregation
+	requestIDMap := map[string]MetricObject{}
+
+	for i := range metricObjects {
```

Comment:
Why not get the objects directly?
```suggestion
	for _, obj := range metricObjects {
```